### PR TITLE
Add map selection to campaign lua

### PR DESCRIPTION
**Pull request by ottml**
Add selection map screen to campaign lua and use it for world campaign.
Add documention.

### CODE REVIEW

File: tests/s25Main/campaign/testCampaignLuaFile.cpp [hunk 1]
```diff
@@ -89,7 +90,7 @@ BOOST_AUTO_TEST_CASE(LoadCampaignDescriptionWithoutTranslation)
         bnw::ofstream file(tmp.get() / "campaign.lua");
 
         file << "campaign ={\
-            version = \"1\",\
+            version = \"2\",\
```
Comment (Flamefire) [inline tests/s25Main/campaign/testCampaignLuaFile.cpp:93]: Similar: Why? Where is this used and why do existing campaigns need to be upgraded if they don't use anything new and the new feature is optional?

File: tests/s25Main/campaign/testCampaignLuaFile.cpp [hunk 2]
```diff
@@ -272,6 +273,96 @@ BOOST_AUTO_TEST_CASE(CampaignDescriptionLoadWithTranslation)
...
+    BOOST_TEST(selectionMap->map.filePath == "<RTTR_GAME>/GFX/PICS/WORLD.LBM");
+    BOOST_TEST(selectionMap->map.index == 0u);
+    BOOST_TEST(selectionMap->missionMapMask.filePath == "<RTTR_GAME>/GFX/PICS/WORLDMSK.LBM");
+    BOOST_TEST(selectionMap->missionMapMask.index == 0u);
+    BOOST_TEST(selectionMap->marker.filePath == "<RTTR_GAME>/DATA/IO/IO.DAT");
+    BOOST_TEST(selectionMap->marker.index == 231u);
+    BOOST_TEST(selectionMap->conquered.filePath == "<RTTR_GAME>/DATA/IO/IO.DAT");
+    BOOST_TEST(selectionMap->conquered.index == 232u);
+    BOOST_TEST(selectionMap->mapOffsetInBackground.x == 64);
+    BOOST_TEST(selectionMap->mapOffsetInBackground.y == 70);
```
Comment (Flamefire) [inline tests/s25Main/campaign/testCampaignLuaFile.cpp:354]: Replace those (here and below) by single checks:

```suggestion
    BOOST_TEST(selectionMap->mapOffsetInBackground == Position(64, 70));
```

You might need to include `PointOutput.hpp`

**Review by Flamefire — CHANGES_REQUESTED** [review]
I like how simple this is :)

Just make sure the changes are backwards compatible: Existing campaigns should continue to work without changes because the new feature is backwards compatible as it only introduces an optional new field. Hence the existing tests and campaigns shouldn't need changes.

File: tests/s25Main/campaign/testCampaignLuaFile.cpp (hunk 1 above)
Comment (ottml) [reply tests/s25Main/campaign/testCampaignLuaFile.cpp:93]: Yes i mixed up the lua version and the campaign version

**Comment by Flamefire** [comment]
> @Flamefire Could it be that you changed the hash? The ci is failing with this error:

I did some work in one repo and didn't realize it might cause some fallout. So yes this is fixed now.

During review I was a bit confused about the `getRequiredLuaVersion` vs `campaign.version`. It seems we require the exact version match of the former which means old lua files won't work. The latter seems to be unused.

IIRC we increase `getRequiredLuaVersion` only when the change is a breaking change. This one isn't so I think we can keep the current version. Or am I missing anything?

File: libs/libGamedata/gameData/CampaignDescription.cpp [hunk 3]
```diff
@@ -34,7 +34,16 @@ CampaignDescription::CampaignDescription(const kaguya::LuaRef& table)
     lua::validatePath(luaFolder);
     mapNames = luaData.getOrDefault("maps", std::vector<std::string>());
 
-    luaData.checkUnused();
+    kaguya::LuaRef entry = table["selectionMap"];
+    if(entry.type() == LUA_TTABLE)
+    {
+        selectionMapData = std::make_optional<SelectionMapInputData>(entry);
+    }
```
Comment (Flamefire) [inline libs/libGamedata/gameData/CampaignDescription.cpp:41]: Why the removal of `checkUnused`? What about 
```suggestion
    selectionMapData = luaData.getOrDefault("selectionMap", std::optional<SelectionMapInputData>());
    luaData.checkUnused();
```

If that doesn't work out of the box we need a `getOptional<SelectionMapInputData>` for `CheckedLuaTable` based on `getOrDefault`

File: libs/libGamedata/gameData/CampaignDescription.cpp [hunk 4]
```diff
@@ -34,7 +34,16 @@ CampaignDescription::CampaignDescription(const kaguya::LuaRef& table)
...
+}
+
+std::optional<SelectionMapInputData> CampaignDescription::getSelectionMapData() const
```
Comment (Flamefire) [inline libs/libGamedata/gameData/CampaignDescription.cpp:44]: No need for copies as sometimes you just want to check IFF there is a value.
```suggestion
const std::optional<SelectionMapInputData>& CampaignDescription::getSelectionMapData() const
```

File: tests/s25Main/campaign/testCampaignLuaFile.cpp [hunk 5]
```diff
@@ -62,6 +64,22 @@ BOOST_AUTO_TEST_CASE(ScriptVersion)
...
+        rttr::test::TmpFolder tmp;
+        {
+            bnw::ofstream file(tmp.get() / "campaign.lua");
+            file << boost::format("function getRequiredLuaVersion()\n return %1%\n end") % 1;
+        }
+
+        CampaignDescription desc;
+        CampaignDataLoader loader(desc, tmp.get());
+        rttr::test::LogAccessor logAcc;
+        BOOST_TEST_REQUIRE(!loader.Load());
```
Comment (Flamefire) [inline tests/s25Main/campaign/testCampaignLuaFile.cpp:78]: > version 2 can load version 1

But asserting it fails to load?

File: tests/s25Main/campaign/testCampaignLuaFile.cpp [hunk 6]
```diff
@@ -102,15 +103,15 @@ BOOST_AUTO_TEST_CASE(LoadCampaignDescriptionWithoutTranslation)
             maps = { \"dessert0.WLD\", \"dessert1.WLD\", \"dessert2.WLD\"}\
         }";
 
-        file << "function getRequiredLuaVersion() return 1 end";
+        file << "function getRequiredLuaVersion() return 2 end";
```
[the remark this answers is not in the document]
Comment (Flamefire) [reply tests/s25Main/campaign/testCampaignLuaFile.cpp:123]: With the new check those changes can go, can't they?

File: tests/s25Main/campaign/testCampaignLuaFile.cpp (hunk 2 above)
Comment (Flamefire) [reply tests/s25Main/campaign/testCampaignLuaFile.cpp:371]: You seem to forgot this ;-)

File: tests/s25Main/campaign/testCampaignLuaFile.cpp [hunk 7]
```diff
@@ -272,6 +290,93 @@ BOOST_AUTO_TEST_CASE(CampaignDescriptionLoadWithTranslation)
...
+    BOOST_TEST(selectionMap->missionMapMask.filePath == "<RTTR_GAME>/GFX/PICS/WORLDMSK.LBM");
+    BOOST_TEST(selectionMap->missionMapMask.index == 0u);
+    BOOST_TEST(selectionMap->marker.filePath == "<RTTR_GAME>/DATA/IO/IO.DAT");
+    BOOST_TEST(selectionMap->marker.index == 231u);
+    BOOST_TEST(selectionMap->conquered.filePath == "<RTTR_GAME>/DATA/IO/IO.DAT");
+    BOOST_TEST(selectionMap->conquered.index == 232u);
+    BOOST_TEST(selectionMap->mapOffsetInBackground.x == 64);
+    BOOST_TEST(selectionMap->mapOffsetInBackground.y == 70);
+    BOOST_TEST(selectionMap->disabledColor == 0x70000000u);
+    BOOST_TEST(selectionMap->missionSelectionInfos.size() == static_cast<size_t>(3));
```
Comment (Flamefire) [inline tests/s25Main/campaign/testCampaignLuaFile.cpp:373]: I guess you were avoiding the signed-unsigned-mismatch warning? More readable:

```suggestion
    BOOST_TEST(selectionMap->missionSelectionInfos.size() == 3u);
```

**Review by Flamefire — CHANGES_REQUESTED** [review]
> The former implementation for loading the campaingn description checked that all keys are read (luaData.checkUnused();) otherwise it failed. So the new format cannot be loaded by already existing builds.

Oh yes, already existing builds...

> We have to increase the version for the world campaign. And change the version check to <= current version. So the new build can also load older versions.

That logic isn't flexible enough either. If we do a breaking change we can not use `<=` although we can worry about that if we do that at one point. Then we can do something like '<major> * 100 + <minor>` and accept anything with the same major version and at least the same minor version.  

Again: For later. We can use the current approach for now

Some remarks left

File: libs/libGamedata/gameData/CampaignDescription.cpp (hunk 3 above)
Comment (ottml) [reply libs/libGamedata/gameData/CampaignDescription.cpp]: added getoptional

File: tests/s25Main/campaign/testCampaignLuaFile.cpp (hunk 5 above)
Comment (ottml) [reply tests/s25Main/campaign/testCampaignLuaFile.cpp]: Extended test

**Comment by Flamefire** [comment]
Thanks a lot!

**Pull request merged**